clang/lib/3C/RewriteUtils.cpp

```diff
@@ -21,41 +21,121 @@
+    MMD->print(DeclStream);
+    assert("Original decl string empty." && !DeclStr.empty());
 
-GlobalVariableGroups::~GlobalVariableGroups() {
-  std::set<std::vector<Decl *> *> VVisited;
-  // Free each of the group.
-  for (auto &CurrV : GlobVarGroups) {
-    // Avoid double free by caching deleted sets.
-    if (VVisited.find(CurrV.second) != VVisited.end())
-      continue;
-    VVisited.insert(CurrV.second);
-    delete (CurrV.second);
+    // Undo the mutation, if applicable.
+    if (VD && Init)
+      VD->setInit(Init);
+
+    return DeclStr;
+  }
+
+  // OK, we have to use mkString.
+  ASTContext &Context = MMD->getASTContext();
+  QualType DType = getTypeOfMultiDeclMember(MMD);
+  if (isPtrOrArrayType(DType)) {
+    CVarOption CVO =
+        (isa<TypedefDecl>(MMD)
+             ? Info.lookupTypedef(PersistentSourceLoc::mkPSL(MMD, Context))
+             : Info.getVariable(MMD, &Context));
+    assert(CVO.hasValue() &&
+           "Missing ConstraintVariable for unchanged multi-decl member");
+    // A function currently can't be a multi-decl member, so this should always
+    // be a PointerVariableConstraint.
+    PVConstraint *PVC = cast<PointerVariableConstraint>(&CVO.getValue());
+    // Currently, we benefit from the ItypesForExtern handling in
+    // mkStringForPVDecl in one very unusual case: an unchanged multi-decl
+    // member with a renamed TagDecl and an existing implicit itype coming from
+    // a bounds annotation will keep the itype and not be changed to a fully
+    // checked type. DeclRewriter::buildItypeDecl will detect the base type
+    // rename and generate the unchecked side using mkString instead of
+    // Decl::print in order to pick up the new name.
+    //
+    // As long as 3C lacks real support for itypes on variables, this is
+    // probably the behavior we want with -itypes-for-extern. If we don't care
+    // about this case, we could alternatively inline the few lines of
+    // mkStringForPVDecl that would still be relevant.
+    return mkStringForPVDecl(MMD, PVC, Info);
   }
-  GlobVarGroups.clear();
+
+  // If the type is not a pointer or array, then it should just equal the base
+  // type except for top-level qualifiers (REVIEW: Can we verify that somehow?),
+  // and it can't have itypes or bounds.
+  // PointerVariableConstraint::extractBaseType doesn't include qualifiers, but
+  // since we know the type is not a pointer, just adding any qualifiers at the
+  // beginning of the string should be correct.
+  std::string QualifierPrefix = DType.getQualifiers().getAsString();
+  if (!QualifierPrefix.empty())
+    QualifierPrefix += " ";
+  // REVIEW: It's awkward to use PointerVariableConstraint::extractBaseType
+  // here, though it seems to work so far.
+  // PointerVariableConstraint::extractBaseType takes a bunch of parameters that
+  // are poorly documented and seem specific to the way it is used by
+  // PointerVariableConstraint. Help me clean that up? Once
```

Comment:
For a start, I believe the `Type`, and `QualType` parameters don't need to be separate. This is more generally an issue in the `PVConstraint` constructor. I think a major improvement to `extractBaseType` would likely come along side substantial clean up of the rest of the constructor.